internal/hud/prompt/prompt.go

```diff
@@ -1,5 +1,128 @@
 package prompt
 
 import (
-	_ "github.com/mattn/go-tty"
+	"context"
+	"fmt"
+
+	tty "github.com/mattn/go-tty"
+	"github.com/pkg/browser"
+
+	"github.com/tilt-dev/tilt/internal/hud"
+	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/pkg/model"
 )
+
+type TerminalInput interface {
+	ReadRune() (rune, error)
+	Close() error
+}
+
+type OpenInput func() (TerminalInput, error)
+
+type OpenURL func(url string) error
+
+func TTYOpen() (TerminalInput, error) {
+	return tty.Open()
+}
+
+func BrowserOpen(url string) error {
+	return browser.OpenURL(url)
+}
+
+type TerminalPrompt struct {
+	openInput OpenInput
+	openURL   OpenURL
+	stdout    hud.Stdout
+	host      model.WebHost
+	url       model.WebURL
+	printed   bool
+}
+
+func NewTerminalPrompt(openInput OpenInput, openURL OpenURL, stdout hud.Stdout, host model.WebHost, url model.WebURL) *TerminalPrompt {
+	return &TerminalPrompt{
+		openInput: openInput,
+		openURL:   openURL,
+		stdout:    stdout,
+		host:      host,
+		url:       url,
+	}
+}
+
+func (p *TerminalPrompt) isEnabled(st store.RStore) bool {
+	state := st.RLockState()
+	defer st.RUnlockState()
+	return state.TerminalMode == store.TerminalModePrompt
+}
+
+func (p *TerminalPrompt) OnChange(ctx context.Context, st store.RStore) {
+	if !p.isEnabled(st) {
+		return
+	}
+
+	if p.printed {
+		return
+	}
+
+	hasBrowserUI := !p.url.Empty()
+	serverStatus := "(without browser UI)"
+	if hasBrowserUI {
+		if p.host == "0.0.0.0" {
+			serverStatus = fmt.Sprintf("on %s (listen on 0.0.0.0)", p.url)
+		} else {
+			serverStatus = fmt.Sprintf("on %s", p.url)
+		}
+	}
+
+	_, _ = fmt.Fprintf(p.stdout, "Tilt started %s\n", serverStatus)
+
+	if hasBrowserUI {
+		_, _ = fmt.Fprintf(p.stdout, "(space) to open the browser\n")
+	}
+
+	// TODO(nick): implement this
+	// _, _ = fmt.Fprintf(p.stdout, "(s) to stream logs\n")
+	_, _ = fmt.Fprintf(p.stdout, "(ctrl-c) to exit\n")
+
+	p.printed = true
+
+	t, err := p.openInput()
+	if err != nil {
+		st.Dispatch(store.ErrorAction{Error: err})
+		return
+	}
+
+	defer func() {
+		_ = t.Close()
+	}()
+
+	keyCh := make(chan rune)
+
+	go func() {
+		for ctx.Err() == nil {
+			r, err := t.ReadRune()
+			if err != nil {
+				st.Dispatch(store.ErrorAction{Error: err})
+				return
+			}
+			keyCh <- r
+		}
+	}()
+
+	for ctx.Err() == nil {
```

Comment:
IIUC, at the moment, this is just holding onto the subscriber's notify goroutine forever, which ~means its OnChange can only ever be called once (which then might surprise someone trying to edit this subscriber). It seems to work, but also seems different from how our other subscribers work, and potentially accidental (like, does this mostly make `p.printed` redundant?)

would it be better to put this for loop in a new goroutine? am I missing something?